docs/proposals/failover/optimize-cluster-failover.md

```diff
@@ -0,0 +1,635 @@
+Next, let's introduce the design solutions for each of the above parts that do not meet the objectives one by one.
+
+### Cluster failure condition judgment
+
+Cluster failure condition judgment involves assessing the status conditions of a faulty cluster object to add various taints to the cluster. This process is referred to as "taint cluster by conditions".
+
+This proposal adds a new API: ClusterTaintPolicy, which can be used to describe what taints should be added to the target cluster when certain cluster status conditions are met. The relevant logic can be implemented by adding a new controller. This controller is responsible for monitoring changes in the cluster status conditions, that is, the `.status.conditions` field of the Cluster object, and then deciding how to modify the taints of the Cluster object according to the ClusterTaintPolicy API defined by the user. When the matching conditions are no longer met, the taints added to the target cluster will be removed.
+
+Key design considerations for ClusterTaintPolicy:
+- To avoid over-design, ClusterTaintPolicy currently mainly provides condition judgment based on `cluster.status.conditions` as input. More input conditions can be added as needed in the future.
+- When making input condition judgments based on `cluster.status.conditions`, in addition to `condition.type` and `condition.status`, time-dimension matching is also required to prevent taint jitter caused by frequent changes in cluster conditions and make the system overly sensitive. However, also to avoid over-design, in the initial version, only simple time-window-based judgments can be provided. For example, if the condition indicating that the cluster is disconnected persists for 5 minutes, add a taint; if the condition indicating that the cluster has re - established connection holds for 3 minutes, remove the taint.
+
+The ClusterTaintPolicy resource is a Cluster scope resource. Its API group is `policy.karmada.io` and the API version is `v1alpha1`.
+
+```go
+// ClusterTaintPolicy defines how Karmada would taint clusters according
+// to conditions on clusters.
+type ClusterTaintPolicy struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec represents the desired behavior of ClusterTaintPolicy.
+    // +required
+    Spec ClusterTaintPolicySpec `json:"spec"`
+}
+
+type ClusterTaintPolicySpec struct {
+    // TargetCluster specifies the clusters that ClusterTaintPolicy needs to pay attention to.
+    // For clusters that meet the MatchConditions, TaintsToAdd will be added.
+    // If targetCluster is not set, any cluster can be selected.
+    // +optional
+    TargetCluster *ClusterAffinity `json:"targetCluster,omitempty"`
+
+    // MatchConditions indicates the conditions to match for triggering
+    // the taint controller to add taints on the cluster object.
+    // The match conditions are ANDed.
+    // If empty, the TaintsToAdd will be added.
+    // +optional
```

Comment:
> If empty, the TaintsToAdd will be added.

If that's the case, the `ClusterTaintPolicy` would essentially become a mechanical tool for adding taints to Clusters, and these taints would never be removed. I don't believe that's the intention of `ClusterTaintPolicy`.

In my opinion, the `MatchConditions` should be `required` instead of `optional`, without a clear match condition, no operation should be done.

Reply:
It makes sense. Let me update the comment.